subsys/power/device_pm.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <kernel.h>
+#include <string.h>
+#include <soc.h>
+#include <device.h>
+#include "policy/pm_policy.h"
+
+#define LOG_LEVEL CONFIG_PM_LOG_LEVEL /* From power module Kconfig */
+#include <logging/log.h>
+LOG_MODULE_DECLARE(power);
+
+#define DEV_RPM_ASYNC	(1 << 0)
+
+static void ipm_work_handler(struct k_work *work)
+{
+	struct device_config *cfg = CONTAINER_OF(work,
+					struct device_config, ipm_work);
+	struct device *dev = device_get_binding(cfg->name);
+	int ret;
+
+	k_sem_take(&dev->config->lock, K_FOREVER);
+
+	/* Incase of suspend request, make sure the device is not busy */
+	if (device_busy_check(dev) &&
+			(cfg->ipm_request == DEVICE_PM_SUSPEND_STATE)) {
+		goto out;
+	}
+
+	ret = device_set_power_state(dev, cfg->ipm_request);
+	if (ret) {
+		LOG_ERR("Set power state(%d) error(%d)\n",
+					ret, cfg->ipm_request);
+	}
+out:
+	k_sem_give(&dev->config->lock);
+}
+
+static int device_ipm_suspend(struct device *dev, u32_t ipm_flags)
+{
+	int ret;
+	u32_t power_state;
+
+	atomic_dec(&dev->config->usage);
+
+	/* add a lock while resuming the device */
+	ret = device_get_power_state(dev, &power_state);
+	if (ret) {
+		LOG_ERR("Get power state error(%d)\n", ret);
+		goto out;
+	}
+
+	/* Return if already suspended */
+	if ((power_state == DEVICE_PM_SUSPEND_STATE) ||
+		(power_state == DEVICE_PM_FORCE_SUSPEND_STATE)) {
+		goto out;
+	}
+
+	/* Return error if device is busy */
+	if (device_busy_check(dev)) {
+		LOG_ERR("dev_ipm: device busy!!\n");
+		return -EBUSY;
+	}
+
+	/* Cancel already queued work */
+	k_delayed_work_cancel(&dev->config->ipm_work);
```

Comment:
I do not know if we can do that without  tracing device state. The driver will see multiple calls to device_set_power_state with the same argument.

Reply:
Ok will rework the logic here.